File: Week_07-字典树-并差集-AVL树/37.解数独.cpp
/*
 * @lc app=leetcode.cn id=37 lang=cpp
 *
 * [37] 解数独
 * theme 高级搜索
 * theme 回溯
 */
// @lc code=start
#include<vector>
using namespace std;
class Solution {
private:
    //行，列，小格内某数字是否已填标记
    bool visitRow[9][9] = { false };
    bool visitCol[9][9] = { false };
    bool visitBox[9][9] = { false };
public:
    void solveSudoku(vector<vector<char>>& board) {
        //先记录表格中的初始状态
        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] != '.'){
                    visitRow[i][ board[i][j] - '1'] = true;
                    visitCol[j][ board[i][j] - '1'] = true;
                    visitBox[ (i / 3) * 3 + j / 3 ][ board[i][j] - '1'] = true;
                }
            }
        }
        backTrace(board, 0, 0);
    }
private:
/**
 * @brief 递归函数
 *        注意其中的return逻辑
 * @param board 
 * @param row 
 * @param col 
 * @return true 
 * @return false 
 */
    bool backTrace(vector<vector<char>>& board, int row, int col) {
//递归step1:递归终止条件
        if (row > 8) return true;
        //如果当前位置空，即没填入数字
        if (board[row][col] == '.') {
//递归step2:处理当前层逻辑        
            for (int i=0; i<9; i++) {
                int boxIndex = (row/3)*3 + col/3;
                if (visitRow[row][i] || visitCol[col][i] || visitBox[boxIndex][i]) continue;
                board[row][col] = i + '1';
                visitRow[row][i] = true;
                visitCol[col][i] = true;
                visitBox[boxIndex][i] = true;
//递归step3:进入到下一层，
//如果下一层返回false， 那就执行step4，回复当前层状态；
//如果下一层返回true（说明下一层或者下一层的下一层...，完美地实现了数独），那么就往上一层也返回true。
                if (backTrace(board, row+(col+1)/9, (col+1)%9)) return true; //如果写成return(backTrace)形式，不能执行下面的恢复当前层状态语句
//递归step4:恢复当前层状态
                board[row][col] = '.';
                visitRow[row][i] = false;
                visitCol[col][i] = false;
                visitBox[boxIndex][i] = false;
            }
            return false;
        }
//递归step3:进入到下一层
        else return backTrace(board, row+(col+1)/9, (col+1)%9);
    }

};
// @lc code=end

File: Week_07-字典树-并差集-AVL树/37.解数独.py
#
# @lc app=leetcode.cn id=37 lang=python
#
# [37] 解数独
#

# @lc code=start
class Solution(object):
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        row = [set(range(1,10)) for _ in range(9)] #行剩余可用数字
        col = [set(range(1,10)) for _ in range(9)] #列剩余可用数字
        box = [set(range(1,10)) for _ in range(9)] #块剩余可用数字
        empty = [] #收集需要填充的位置
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.':  #更新可用数字
                    val = int(board[i][j])
                    row[i].remove(val)
                    col[j].remove(val)
                    box[(i//3)*3+j//3].remove(val)
                else:
                    empty.append((i,j))  #统计要填充的位置
        def backtrack(iter):
            if iter == len(empty):  #处理完要填充的位置就找到了答案
                return True
            i, j = empty[iter]
            for val in row[i] & col[j] & box[(i//3)*3+j//3]:
                row[i].remove(val)
                col[j].remove(val)
                box[(i//3)*3+j//3].remove(val)
                board[i][j] = str(val)
                if backtrack(iter+1): return True  
                row[i].add(val)
                col[j].add(val)
                box[(i//3)*3+j//3].add(val)
            return False
        backtrack(0)
        


# @lc code=end

